pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java

```diff
@@ -256,4 +257,61 @@ private static boolean verifyIntegerModifiers(String[] modifiers) {
             throw new RuntimeException("Failed connecting to Hive MetaStore service: " + cause.getMessage(), cause);
         }
     }
+
+
+    /**
+     * Converts HAWQ type to hive type.
+     * @see EnumHiveToHawqType For supported mappings
+     * @param type HAWQ data type
+     * @param name field name
+     * @return Hive type
+     * @throws UnsupportedTypeException if type is not supported
+     */
+    public static String toCompatibleHiveType(DataType type) {
+
+        EnumHiveToHawqType hiveToHawqType = EnumHiveToHawqType.getCompatibleHiveToHawqType(type);
+        return hiveToHawqType.getTypeName();
+    }
+
+
+
+    /**
+     * Validates whether given HAWQ and Hive data types are compatible
+     *
+     * @param hawqDataType HAWQ data type
+     * @param hawqTypeMods HAWQ type modifiers
+     * @param hiveType full Hive type, i.e. decimal(10,2)
+     * @param hawqColumnName Hive column name
+     * @throws UnsupportedTypeException if types are incompatible
+     */
+    public static void validateTypeCompatible(DataType hawqDataType, Integer[] hawqTypeMods, String hiveType, String hawqColumnName) {
+
+        EnumHiveToHawqType hiveToHawqType = EnumHiveToHawqType.getHiveToHawqType(hiveType);
+        EnumHawqType expectedHawqType = hiveToHawqType.getHawqType();
+
+        if (!expectedHawqType.getDataType().equals(hawqDataType)) {
+            throw new UnsupportedTypeException("Invalid definition for column " + hawqColumnName
+                                    +  ": expected HAWQ type " + expectedHawqType.getDataType() +
+                    ", actual HAWQ type " + hawqDataType);
+        }
+
+        switch (hawqDataType) {
+            case NUMERIC:
+            case VARCHAR:
+            case BPCHAR:
+                if (hawqTypeMods != null && hawqTypeMods.length > 0) {
+                    Integer[] hiveTypeModifiers = EnumHiveToHawqType
+                            .extractModifiers(hiveType);
+                    for (int i = 0; i < hiveTypeModifiers.length; i++) {
+                        if (hawqTypeMods[i] < hiveTypeModifiers[i])
+                            throw new UnsupportedTypeException(
+                                    "Invalid definition for column " + hawqColumnName
+                                            + ": modifiers are not compatible, "
+                                            + Arrays.toString(hiveTypeModifiers) + ", "
+                                            + Arrays.toString(hawqTypeMods));
+                    }
+                }
+                break;
```

Comment:
Right now both cases are valid:

Hive type smallint, HAWQ type smallint, Hive type tinyint, HAWQ type smallint.
They are covered by tests in https://github.com/apache/incubator-hawq/pull/873/files#diff-7d6d87697d9432f93698702c36eda0d3R187